app/src/main/java/ch/epfl/sdp/artificial_intelligence/Enemy.java

```diff
@@ -0,0 +1,155 @@
+package ch.epfl.sdp.artificial_intelligence;
+
+import java.util.List;
+
+import ch.epfl.sdp.Player;
+
+public class Enemy extends MovingEntity {
+    private Behaviour behaviour;
+    private List<Player> players; // For now I use a list of players, but it could be nice to have
+    // a static manager of players.
+    private int damage;
+    private float dps; // damage per second
+    private float detectionDistance;
+    private int timeAttack;
+    private int timeWandering;
+    private LocalBounds patrolBounds;
+    private Boundable maxBounds;
+    private boolean waiting;
+
+    public Enemy(List<Player> players, int damage, float dps, float detectionDistance, LocalBounds patrolBounds, Boundable maxBounds) {
+        this.damage = damage;
+        this.dps = dps;
+        this.detectionDistance = detectionDistance;
+        this.players = players;
+        behaviour = Behaviour.WAIT;
+        timeAttack = 100; // Needs calibration
+        timeWandering = 1000;
+        this.patrolBounds = patrolBounds;
+        this.maxBounds = maxBounds;
+        this.waiting = false;
+    }
+
+    private void behave() {
+        switch (behaviour) {
+            case ATTACK:
+                attack();
+                break;
+            case CHASE:
+                chase();
+                break;
+            case PATROL:
+                patrol();
+                break;
+            case WAIT:
+                _wait();
+                break;
+            case WANDER:
+                wander();
+                break;
+        }
+    }
+
+    private void _wait() {
+        if (!waiting) {
+            super.setMoving(true);
+            behaviour = Behaviour.WANDER;
+        }
+    }
+
+    public void setWaiting(boolean waiting) {
+        this.waiting = waiting;
+    }
+
+    private void attack() {
+        if (timeAttack <= 0) {
+            Player target = selectClosestPlayer();
+
+            if (target != null) {
+                target.setHealthPoints(target.getHealthPoints() - damage * dps);
+            }
+
+            timeAttack = 100;
+        } else {
+            timeAttack -= 1;
+        }
+
+        if (waiting) {
+            super.setMoving(false);
+            behaviour = Behaviour.WAIT;
+        }
+    }
+
+    private Player selectClosestPlayer() {
+        Player target = null;
+        float minDistance = Float.MAX_VALUE;
+
+        for (Player player : players) {
+            if (player.getPosition().toCartesian().distanceFrom(getPosition()) < minDistance && player.isAlive()) {
+                target = player;
+            }
+        }
+
+        return target;
+    }
+
+    private void chase() {
+        if (playerDetected()) {
+            Player target = selectClosestPlayer();
```

Comment:
playerDetected already uses selectClosestPlayer, isn't this call redundant?

Reply:
Maybe you can directly detect the player and give it as argument to a method isPlayerDetected(Player player) which will tell if he is inside the radius)